decentraland/unity-renderer
# refactor: WSSController now is a production tool, and it was split to its own class

## What does this PR change?

Fix #585

ADR for this refactor: WIP

After this refactor the life cycle dependency was this: 
![WSSControllerRefactor](https://user-images.githubusercontent.com/12563266/130438310-ad41b1e5-bec2-47e5-b3e8-fe753a930704.png)

And now, this PR's does the following:
![WSSControllerRefactor (2)](https://user-images.githubusercontent.com/12563266/130438454-1d49b855-fc78-46d9-9907-871a10a7bc0a.png)

- I moved the WebSocket behaviour from `WSSController` to a new `WebSocketCommunication` class, an interface to `IKernelCommunication`.
- `EntryPoint_World` was renamed to `KernelBridgeCommunication`, and it has an interface with `IKernelCommunication`
- The old `WSSController` was renamed to `DebugParameters`, and its only responsibility is to open the Web Browser to execute the `Kernel`.
- The messages handling in the `WebSocketCommunication` was changed to only use `SendMessages`. I added a function to see which bridge is needed from each message (because the `Kernel` doesn't send to WebSocket which Bridge needs)
- Now, we cannot choose which port the WebSocket will communicate (because it's a version to go to production in `Desktop`), so now we look for available ports between 5000 to 6000. Maybe in the future, we can add a way to overwrite this if it's needed.
- I added some DataStore variables to know if the communication it's ready (WebSocket Server is listening) and when the communication is established (the `Kernel` connected with the `renderer`), this last one I will use in `Desktop`.
- When we send a Message to the `Kernel` from the `renderer`, and there is no function assigned to send the message, we queue it and send it when the connection is established to avoid race conditions.

## How to test the changes?

1. Checkout this branch in your local computer
2. Open Unity, in the `InitialScene` go to `DebugParameters` and set `custom URL` to the following: `?DEBUG_MESSAGES&renderer-branch=refactor/wsscontroller&`
3. Touch `Play`
4. Check if everything goes well

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
